hbase-it/src/test/java/org/apache/hadoop/hbase/ShellExecCoproc.java

```diff
@@ -0,0 +1,149 @@
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.coprocessor.MasterCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.RegionServerCoprocessor;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecRequest;
+import org.apache.hadoop.hbase.coprocessor.protobuf.generated.ShellExecEndpoint.ShellExecResponse;
+import org.apache.hadoop.hbase.ipc.CoprocessorRpcUtils;
+import org.apache.hadoop.util.Shell;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Receives shell commands from the client and executes them blindly.
+ */
+@InterfaceAudience.Private
+public class ShellExecCoproc
+  extends ShellExecEndpoint.ShellExecService
+  implements MasterCoprocessor, RegionServerCoprocessor {
+  private static final Logger LOG = LoggerFactory.getLogger(ShellExecCoproc.class);
+
+  public static final String ASYNC_DELAY_KEY = "hbase.it.shellexeccoproc.async.delay";
+  public static final long DEFAULT_ASYNC_DELAY = 1_000;
+
+  private final ExecutorService workerPool;
+  private Configuration conf;
+
+  public ShellExecCoproc() {
+    workerPool = Executors.newSingleThreadExecutor(
+      new ThreadFactoryBuilder()
+        .setNameFormat(ShellExecCoproc.class.getSimpleName() + "-{}")
+        .setDaemon(true)
+        .setUncaughtExceptionHandler((t, e) -> LOG.warn("Thread {} threw", t, e))
+        .build());
+  }
+
+  @Override
+  public Iterable<Service> getServices() {
+    return Collections.singletonList(this);
+  }
+
+  @Override
+  public void start(CoprocessorEnvironment env) {
+    conf = env.getConfiguration();
+  }
+
+  @Override
+  public void shellExec(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, false);
+
+    ShellExecResponse.Builder builder = ShellExecResponse.newBuilder();
+    try {
+      builder = doExec(shell, builder);
+    } catch (IOException e) {
+      LOG.info("Failure launching process", e);
+      CoprocessorRpcUtils.setControllerException(controller, e);
+    }
+
+    done.run(builder.build());
+  }
+
+  @Override
+  public void shellExecAsync(
+    final RpcController controller,
+    final ShellExecRequest request,
+    final RpcCallback<ShellExecResponse> done
+  ) {
+    final Shell.ShellCommandExecutor shell = prepareShell(request, true);
+    final long sleepDuration = conf.getLong(ASYNC_DELAY_KEY, DEFAULT_ASYNC_DELAY);
+    workerPool.submit(() -> {
+      try {
+        // sleep first so that the RPC can ACK.
```

Comment:
Don't fully understand this, is there a race?

Reply:
Yes. The primary use-case is for the process to call `kill -9` on itself. If we hold up the RPC waiting for a response, it never comes because the process dies. Then the client retry mechanism sends the RPC again, this time to the new active master, until we run out of masters.

Our client doesn't let us say "this RPC is intended for this one host only." So we get the above behavior. By making the invocation a set-and-forget, with a delay on execution, I'm giving the RPC a change to ACK back to the client before the host terminates itself.

Reply:
Ah, the kill -9 for chaos monkey.. now I see it. The method name "shellExecAsync()" gave an impression that it is a generic shell exec utility. Also I think you added the "async" version just so that it acks the RPC and sends a response back. I think we should do one of the following

1. Since we already have a blocking version and async version is only for kill, a special command (IIUC) rename it to make it more self explanatory?
2. If you don't agree with above, add some comments why this async version exists?

Reply:
I've adjusted names and added comments. Please let me know if the confusion persists.